magpie/api/management/network_node/network_node_views.py

```diff
@@ -0,0 +1,89 @@
+from pyramid.httpexceptions import (
+    HTTPBadRequest,
+    HTTPConflict,
+    HTTPNotFound,
+    HTTPOk,
+    HTTPCreated,
+    HTTPInternalServerError
+)
+
+from pyramid.view import view_config
+
+from magpie import models
+from magpie.api import exception as ax
+from magpie.api import requests as ar
+from magpie.api import schemas as s
+from magpie.api.management.network_node.network_node_utils import create_network_node, delete_network_node
+from magpie.constants import get_constant
+
+
+if get_constant("MAGPIE_NETWORK_MODE", settings_name="magpie.network_mode"):
+    @s.NetworkNodesAPI.get(tags=[s.NetworkNodeTag], response_schemas=s.NetworkNodes_GET_responses)
+    @view_config(route_name=s.NetworkNodesAPI.name, request_method="GET")
+    def get_network_nodes_view(request):
+        nodes = [{"name": n.name, "url": n.url} for n in request.db.query(models.NetworkNode).all()]
+        return ax.valid_http(http_success=HTTPOk, detail=s.NetworkNode_GET_OkResponseSchema.description,
+                             content={"nodes": nodes})
+
+
+    @s.NetworkNodeAPI.get(tags=[s.NetworkNodeTag], response_schemas=s.NetworkNode_GET_responses)
+    @view_config(route_name=s.NetworkNodeAPI.name, request_method="GET")
+    def get_network_node_view(request):
+        node_name = ar.get_value_matchdict_checked(request, "node")
+        node = ax.evaluate_call(
+            lambda: request.db.query(models.NetworkNode).filter(models.NetworkNode.name == node_name).one(),
+            http_error=HTTPNotFound,
+            msg_on_fail=s.NetworkNode_GET_NotFoundResponseSchema.description)
+        return ax.valid_http(http_success=HTTPOk, detail=s.NetworkNode_GET_OkResponseSchema.description,
+                             content={"name": node.name, "url": node.url})
+
+
+    @s.NetworkNodesAPI.post(schema=s.NetworkNode_POST_RequestBodySchema, tags=[s.NetworkNodeTag],
+                            response_schemas=s.NetworkNodes_POST_responses)
+    @view_config(route_name=s.NetworkNodesAPI.name, request_method="POST")
+    def create_network_node_view(request):
+        node_name = request.POST.get("name")
+        node_url = request.POST.get("url")
+        ax.verify_param(all([node_name, node_url]), is_true=True,
+                        http_error=HTTPBadRequest, msg_on_fail=s.BadRequestResponseSchema.description)
+        ax.evaluate_call(lambda: create_network_node(request, node_name, node_url),
+                         http_error=HTTPConflict,
+                         msg_on_fail=s.NetworkNodes_POST_ConflictResponseSchema.description)
+        return ax.valid_http(http_success=HTTPCreated, detail=s.NetworkNode_GET_OkResponseSchema.description)
```

Comment:
Might need more than this. 

For now, any admin could register a node, which could be updated/deleted/recreated at some other point by any other admin, without any handshake validation of the node and associated tokens generated for it. If one node in the network gets hacked, that could allow a malicious user to generate tokens over the entire network.

Similar to when registering an external auth provider, there should be some kind of client ID/secret pair that is generated, and which the user must register back on their side to sign the tokens for that new node. This way, when a token is generated for a given node and sent to it, that node would have some ID to validate the token against and ensure it comes from a trusted location.

Reply:
Ok a few questions about this...

> For now, any admin could register a node, which could be updated/deleted/recreated at some other point by any other admin

Do you mean an admin for the same node or an admin for a different node? If you're talking about an admin for node A changing stuff of node B, I don't know how they would do that unless they had an account on both?

> If one node in the network gets hacked, that could allow a malicious user to generate tokens over the entire network.

Yes, I guess that if a node got hacked they could validate a token and return any user name they wanted. Not that that isn't bad and requires some thought, but I think that even with Oauth2, if the provider is compromised they can send any response they want anyway. Is this scenario different from that?

> This way, when a token is generated for a given node and sent to it, that node would have some ID to validate the token against and ensure it comes from a trusted location.

Right now, the JWT is validated to ensure that the token was issued by the right node. We could add another check to ensure that it was sent by one of the expected location(s). I think that's a great idea. Would that be something that using Oauth and authomatic would automatically get us?

Reply:
> Do you mean an admin for the same node or an admin for a different node? 

An admin of the node where the POST/PUT/DELETE requests are done. This is because of the default security check of Magpie that blocks requests from non-admin users. 

> Yes, I guess that if a node got hacked they could validate a token and return any user name they wanted. Not that that isn't bad and requires some thought, but I think that even with Oauth2, if the provider is compromised they can send any response they want anyway. Is this scenario different from that?

It wouldn't stop someone from sending requests from within the hacked node, but if an admin-user is compromised, it makes it slightly more complicated for someone to try to impersonate a fake Magpie node, because the other nodes that they could try to send token validations to would not have registered/handshake that fake node. It would also be easier to respond to some breach by revoking the hacked node client ID/secret. 

> Right now, the JWT is validated to ensure that the token was issued by the right node. We could add another check to ensure that it was sent by one of the expected location(s). I think that's a great idea. Would that be something that using Oauth and authomatic would automatically get us?

Yes. This is something like what I had in mind. However, the "location" doesn't need to be a URL, because that can be spoofed as well with requests. Doesn't hurt to have more validation layers though.
Yes, most of the features should be fairly well supported. It is used here for example https://github.com/Ouranosinc/Magpie/blob/master/magpie/security.py#L171-L190